File: 0063-unique-paths-ii/0063-unique-paths-ii.py
class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        n=len(obstacleGrid)
        m=len(obstacleGrid[0])
        prev=[0]*(m+1)
        cur=[0]*(m+1)
        for ind1 in range(1,n+1):
            for ind2 in range(1,m+1):
                if obstacleGrid[ind1-1][ind2-1]==1:
                    cur[ind2]=0
                else:
                    if ind1==1 and ind2==1:
                        cur[ind2]=1
                    else:
                        up,down=0,0
                        if ind1-1>=0:
                            up=prev[ind2]
                        if ind2-1>=0:
                            down=cur[ind2-1]
                        cur[ind2]=up+down
            prev=cur
        return prev[m]
                
        
        
        
        
#         def solve(ind1,ind2,obstacleGrid):
#             if obstacleGrid[ind1-1][ind2-1]==1:
#                 return 0
#             if dp[ind1][ind2]!=-1:
#                 return dp[ind1][ind2]
#             else:
#                 if ind1==0 or ind2==0:
#                     return 0
#                 if ind1==1 and ind2==1:
#                     return 1
#                 else:
#                     left=0
#                     down=0
#                     if ind1-1>=0:
#                         up=solve(ind1-1,ind2,obstacleGrid)
#                     if ind2-1>=0:
#                         down=solve(ind1,ind2-1,obstacleGrid)
#                     dp[ind1][ind2]=up+down
#                     return dp[ind1][ind2]
#         return solve(n,m,obstacleGrid)
        
        

File: 0063-unique-paths-ii/0063-unique-paths-ii.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {
        int m=obstacleGrid.size();
        int n=obstacleGrid[0].size();
        int dp[m][n];
        for(int i=0;i<m;i++){
            for(int j=0;j<n;j++){
                if(obstacleGrid[i][j]==1){
                    dp[i][j]=0;
                }
                else if(i==0 and j==0)dp[i][j]=1;
                else{
                    int down=0;
                    int left=0;
                    if(i-1>=0)down=obstacleGrid[i][j]+dp[i-1][j];
                    if(j-1>=0)left=obstacleGrid[i][j]+dp[i][j-1];
                    dp[i][j]=down+left;
                }
            }
        }
        return dp[m-1][n-1];
    }
};